File: framework/languages/dotnet/e2e/RegistryMessaging/Client/Support/DynamicClusterLauncher.cs
using System.Diagnostics;
using System.Net;
using System.Net.Sockets;
using Zlink.HttpClient;

namespace RegistryMessaging.Client.Support;

internal sealed class DynamicClusterLauncher(string providerProject, string logDir) : IAsyncDisposable
{
    private readonly List<DynamicProcess> _processes = [];

    public string RegistryRouterEndpoint { get; private set; } = "";

    public async ValueTask DisposeAsync()
    {
        for (var i = _processes.Count - 1; i >= 0; i--) await _processes[i].StopAsync();

        _processes.Clear();
    }

    public static async Task<DynamicClusterLauncher> StartAsync(ClientOptions options, string scenarioName)
    {
        var launcher = new DynamicClusterLauncher(options.ProviderProject, options.LogDir);
        try
        {
            launcher.RegistryRouterEndpoint = PickEndpoint();
            var registryHttp = PickHttpUrl();
            var registry = launcher.StartServer(
                $"{scenarioName}-registry",
                options.RegistryProject,
                [
                    "--rid", $"{scenarioName}-registry",
                    "--http-url", registryHttp,
                    "--registry-pub-endpoint", PickEndpoint(),
                    "--registry-router-endpoint", launcher.RegistryRouterEndpoint,
                    "--log-dir", options.LogDir
                ],
                registryHttp,
                null);
            await registry.WaitReadyAsync();
            return launcher;
        }
        catch
        {
            await launcher.DisposeAsync();
            throw;
        }
    }

    public async Task<DynamicProvider> StartProviderAsync(string name, string rid, int weight = 100)
    {
        var httpUrl = PickHttpUrl();
        var channelEndpoint = PickEndpoint();
        var process = StartServer(
            name,
            providerProject,
            [
                "--rid", rid,
                "--http-url", httpUrl,
                "--registry-router-endpoint", RegistryRouterEndpoint,
                "--channel-endpoint", channelEndpoint,
                "--route-endpoint", PickEndpoint(),
                "--weight", weight.ToString(),
                "--evidence-file", Path.Combine(logDir, $"{name}.evidence.log"),
                "--log-dir", logDir
            ],
            httpUrl,
            channelEndpoint);
        await process.WaitReadyAsync();
        return new DynamicProvider(process, httpUrl, process.RequireChannelEndpoint());
    }

    public async Task StopAsync(DynamicProvider provider)
    {
        await provider.Process.StopAsync();
        _processes.Remove(provider.Process);
    }

    private DynamicProcess StartServer(
        string name,
        string projectPath,
        IReadOnlyList<string> arguments,
        string httpUrl,
        string? channelEndpoint)
    {
        var startInfo = new ProcessStartInfo
        {
            FileName = "dotnet",
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false
        };
        startInfo.ArgumentList.Add("run");
        startInfo.ArgumentList.Add("--project");
        startInfo.ArgumentList.Add(projectPath);
        startInfo.ArgumentList.Add("--");
        foreach (var argument in arguments) startInfo.ArgumentList.Add(argument);

        var process = Process.Start(startInfo)
                      ?? throw new InvalidOperationException($"Failed to start {name}.");
        _ = CopyToFileAsync(process.StandardOutput, Path.Combine(logDir, $"{name}.stdout.log"));
        _ = CopyToFileAsync(process.StandardError, Path.Combine(logDir, $"{name}.stderr.log"));
        var dynamicProcess = new DynamicProcess(process, httpUrl, channelEndpoint);
        _processes.Add(dynamicProcess);
        return dynamicProcess;
    }

    private static async Task CopyToFileAsync(StreamReader reader, string path)
    {
        await using var stream = File.Open(path, FileMode.Create, FileAccess.Write, FileShare.Read);
        await using var writer = new StreamWriter(stream);
        while (await reader.ReadLineAsync() is { } line)
        {
            await writer.WriteLineAsync(line);
            await writer.FlushAsync();
        }
    }

    private static string PickEndpoint()
    {
        return $"tcp://127.0.0.1:{PickPort()}";
    }

    private static string PickHttpUrl()
    {
        return $"http://127.0.0.1:{PickPort()}";
    }

    private static int PickPort()
    {
        using var listener = new TcpListener(IPAddress.Loopback, 0);
        listener.Start();
        return ((IPEndPoint)listener.LocalEndpoint).Port;
    }
}

internal sealed record DynamicProvider(DynamicProcess Process, string HttpUrl, string ChannelEndpoint);

internal sealed class DynamicProcess(Process process, string httpUrl, string? channelEndpoint)
{
    private bool _disposed;

    public string HttpUrl { get; } = httpUrl;

    public string? ChannelEndpoint { get; } = channelEndpoint;

    public string RequireChannelEndpoint()
    {
        return ChannelEndpoint ??
               throw new InvalidOperationException("This process does not expose a channel endpoint.");
    }

    public async Task WaitReadyAsync()
    {
        using var client = new HttpClient();
        for (var i = 0; i < 120; i++)
        {
            if (process.HasExited)
                throw new InvalidOperationException($"Process exited before readiness: {process.ExitCode}.");

            try
            {
                using var response = await client.GetAsync($"{HttpUrl}/health");
                if (response.IsSuccessStatusCode) return;
            }
            catch
            {
            }

            await Task.Delay(250);
        }

        throw new TimeoutException($"Process did not become ready: {HttpUrl}.");
    }

    public async Task StopAsync()
    {
        if (_disposed) return;

        _disposed = true;
        if (!process.HasExited)
            try
            {
                using var client = ZLinkHttpClient.Create(HttpUrl)
                    .Timeout(TimeSpan.FromSeconds(5))
                    .Build();
                await client.Post("/shutdown").SubmitRawAsync();
            }
            catch
            {
                if (!process.HasExited) process.Kill(true);
            }

        await process.WaitForExitAsync();
        process.Dispose();
    }
}

File: framework/languages/node/e2e/RegistryMessaging/Client/Support/dynamic-cluster-launcher.ts
import fs from 'node:fs';
import net from 'node:net';
import path from 'node:path';
import { spawn } from 'node:child_process';
import type { ChildProcess } from 'node:child_process';
import type { ClientOptions } from './client-options';

export interface DynamicProvider {
  readonly process: DynamicProcess;
  readonly httpUrl: string;
  readonly channelEndpoint: string;
}

export class DynamicClusterLauncher {
  private readonly processes: DynamicProcess[] = [];
  private constructor(
    private readonly locationProbeMain: string,
    private readonly providerMain: string,
    private readonly logDir: string,
    private readonly redisEndpoint: string,
    private readonly redisKeyPrefix: string
  ) {}

  static async start(options: ClientOptions, scenarioName: string): Promise<DynamicClusterLauncher> {
    const launcher = new DynamicClusterLauncher(
      options.locationProbeMain,
      options.providerMain,
      options.logDir,
      options.redisEndpoint,
      `${options.redisKeyPrefix}:${scenarioName}`
    );
    try {
      const locationProbeHttp = await pickHttpUrl();
      const locationProbe = launcher.startServer(
        `${scenarioName}-location-probe`,
        options.locationProbeMain,
        {
          rid: `${scenarioName}-location-probe`, httpUrl: locationProbeHttp,
          redisEndpoint: options.redisEndpoint, redisKeyPrefix: `${options.redisKeyPrefix}:${scenarioName}`,
          logDir: options.logDir
        },
        locationProbeHttp,
        undefined
      );
      await locationProbe.waitReady();
      return launcher;
    } catch (error) {
      await launcher.close();
      throw error;
    }
  }

  async startProvider(name: string, rid: string, weight = 100): Promise<DynamicProvider> {
    const httpUrl = await pickHttpUrl();
    const channelEndpoint = await pickEndpoint();
    let process: DynamicProcess | undefined;
    try {
      process = this.startServer(
        name,
        this.providerMain,
        {
          rid, httpUrl, redisEndpoint: this.redisEndpoint, redisKeyPrefix: this.redisKeyPrefix,
          channelEndpoint, routeEndpoint: await pickEndpoint(), routePeers: [], weight,
          maxMessageSize: 0, evidenceFile: path.join(this.logDir, `${name}.evidence.log`), logDir: this.logDir
        },
        httpUrl,
        channelEndpoint
      );
      await process.waitReady();
      return { process, httpUrl, channelEndpoint };
    } catch (error) {
      if (process !== undefined) {
        await process.stop();
        const index = this.processes.indexOf(process);
        if (index >= 0) {
          this.processes.splice(index, 1);
        }
      }
      throw error;
    }
  }

  async stop(provider: DynamicProvider): Promise<void> {
    await provider.process.stop();
    const index = this.processes.indexOf(provider.process);
    if (index >= 0) {
      this.processes.splice(index, 1);
    }
  }

  async close(): Promise<void> {
    for (let i = this.processes.length - 1; i >= 0; i -= 1) {
      await this.processes[i].stop();
    }
    this.processes.length = 0;
  }

  private startServer(name: string, mainPath: string, e2e: Record<string, unknown>, httpUrl: string, channelEndpoint?: string): DynamicProcess {
    const config = path.join(this.logDir, `${name}.config.json`);
    fs.mkdirSync(this.logDir, { recursive: true });
    fs.writeFileSync(config, `${JSON.stringify({ e2e }, null, 2)}\n`, { mode: 0o600 });
    const child = spawn(process.execPath, [mainPath, '--config', config], { stdio: ['ignore', 'pipe', 'pipe'] });
    fs.mkdirSync(this.logDir, { recursive: true });
    child.stdout?.pipe(fs.createWriteStream(path.join(this.logDir, `${name}.stdout.log`)));
    child.stderr?.pipe(fs.createWriteStream(path.join(this.logDir, `${name}.stderr.log`)));
    const dynamicProcess = new DynamicProcess(child, httpUrl, channelEndpoint);
    this.processes.push(dynamicProcess);
    return dynamicProcess;
  }
}

export class DynamicProcess {
  constructor(
    private readonly process: ChildProcess,
    readonly httpUrl: string,
    readonly channelEndpoint?: string
  ) {}

  async waitReady(): Promise<void> {
    for (let i = 0; i < 120; i += 1) {
      if (this.process.exitCode !== null) {
        throw new Error(`Process exited before readiness: ${this.process.exitCode}`);
      }
      try {
        const response = await fetch(`${this.httpUrl}/health`);
        if (response.ok) {
          return;
        }
      } catch {
      }
      await new Promise((resolve) => setTimeout(resolve, 250));
    }
    throw new Error(`Process did not become ready: ${this.httpUrl}`);
  }

  async stop(): Promise<void> {
    if (this.process.exitCode !== null) {
      return;
    }
    const exited = new Promise<void>((resolve) => {
      this.process.once('exit', () => resolve());
    });
    if (this.process.exitCode !== null) {
      return;
    }
    const killer = setTimeout(() => {
      if (this.process.exitCode === null) {
        this.process.kill('SIGKILL');
      }
    }, 5000);
    try {
      await Promise.race([
        fetch(`${this.httpUrl}/shutdown`, { method: 'POST' }),
        new Promise((_, reject) => setTimeout(() => reject(new Error('shutdown request timed out')), 1000))
      ]);
    } catch {
      if (this.process.exitCode === null) {
        this.process.kill('SIGTERM');
      }
    }
    await exited.finally(() => clearTimeout(killer));
  }
}

async function pickEndpoint(): Promise<string> {
  return `tcp://127.0.0.1:${await pickPort()}`;
}

async function pickHttpUrl(): Promise<string> {
  return `http://127.0.0.1:${await pickPort()}`;
}

function pickPort(): Promise<number> {
  return new Promise((resolve, reject) => {
    const server = net.createServer();
    server.once('error', reject);
    server.listen(0, '127.0.0.1', () => {
      const address = server.address();
      const port = typeof address === 'object' && address !== null ? address.port : 0;
      server.close(() => resolve(port));
    });
  });
}
